manager/network/cni/cni.go

```diff
@@ -0,0 +1,84 @@
+package cni
+
+import (
+	"github.com/containernetworking/cni/libcni"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/allocator/cniallocator"
+	"github.com/docker/swarmkit/manager/network"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+type cni struct{}
+
+// New produces a fresh CNI Network Model
+func New() network.Model {
+	return &cni{}
+}
+
+func (nm *cni) NewAllocator() (network.Allocator, error) {
+	return cniallocator.New()
+}
+
+func (nm *cni) SupportsIngressNetwork() bool {
+	return false
+}
+
+func parseCNIspec(spec *api.NetworkSpec) (*libcni.NetworkConfig, error) {
+	// This is rather similar to cniConfig in the containerd executor...
+	cniConfig, ok := spec.DriverConfig.Options["config"]
+	if !ok {
+		return nil, grpc.Errorf(codes.InvalidArgument, "CNI network has no config")
+	}
+
+	cni, err := libcni.ConfFromBytes([]byte(cniConfig))
+	if err != nil {
+		return nil, grpc.Errorf(codes.InvalidArgument, "Failed to parse CNI config: %s", err)
+	}
+	return cni, nil
+}
+
+func (nm *cni) SetDefaults(spec *api.NetworkSpec) error {
+	if spec.DriverConfig == nil || spec.DriverConfig.Name != "cni" {
+		return nil
+	}
+
+	cni, err := parseCNIspec(spec)
+	if err != nil {
+		return err
+	}
+
+	if spec.Annotations.Name == "" {
+		spec.Annotations.Name = cni.Network.Name
```

Comment:
Just noticed this bit. It's sort of problematic - controlapi shouldn't modify the spec, which should only be user data. I can't think of any horrible implications from this particular statement, but we've worked pretty hard to keep a clear separation between user-controlled data and fields that are modifiable by swarmkit, and I wouldn't want to introduce an exception without a very good reason. Among other things, we're thinking of adding a mechanism for signing specs before they're submitted to swarmkit, in the future.

Could we remove `SetDefaults` and instead dynamically resolve an empty`spec.Annotations.Name` field?